### PR TITLE
Add support for DatetimeTZDtype and tz_localize

### PR DESCRIPTION
**Pull request by shwina**
## Description

TBD

Quick benchmark:

```python
psr = pd.Series(pd.date_range("1970-01-01", "1980-01-01", freq="1T"))
sr = cudf.from_pandas(psr)

In [6]: %timeit psr.dt.tz_localize("America/New_York", ambiguous="NaT", nonexistent="NaT")
1.55 s ± 9.91 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: %timeit sr.dt.tz_localize("America/New_York", ambiguous="NaT", nonexistent="NaT")
9.98 ms ± 246 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/index.py
Comment (mroeschke) [inline python/cudf/cudf/core/index.py:2369]: As a pandas 2.0 note, would be good to keep "s", "ms", or "us" resolution here

File: python/cudf/cudf/core/column/datetime.py
Comment (mroeschke) [inline python/cudf/cudf/core/column/datetime.py:568]: I think you'll need to use `pa.assume_timezone` here. IIRC `cast` is equivalent to "tz_convert"  in pandas: https://arrow.apache.org/docs/python/generated/pyarrow.compute.assume_timezone.html

File: python/cudf/cudf/core/_internals/timezones.py
Comment (mroeschke) [inline python/cudf/cudf/core/_internals/timezones.py:160]: Are you planning on supporting `tz_localize(None)` in this PR? In pandas, it removes the timezone info for tz-aware data returning local timestamps

File: python/cudf/cudf/core/index.py
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py:2369]: Since this is currently targeting `branch-23.06` which doesn't have pandas-2.0 support this change seems okay, what @mroeschke suggesting is taken care of in `pandas_2.0_feature_branch`, so perhaps a note or todo here would be useful while I sync these changes with pandas-2.0 support.

File: python/cudf/cudf/_lib/column.pyx
Comment (bdice) [inline python/cudf/cudf/_lib/column.pyx:322]: Let's use `col` or `self` consistently in the second line. Favoring `col` here:
```suggestion
            data_dtype = col.dtype
```

File: python/cudf/cudf/core/_internals/timezones.py
Comment (bdice) [inline python/cudf/cudf/core/_internals/timezones.py:101]: Please document the parameters and two return values. I had to scroll to see `return ambiguous, nonexistent` before I understood what this was meant to do.

File: python/cudf/cudf/core/column/datetime.py
Comment (bdice) [inline python/cudf/cudf/core/column/datetime.py:578]:
```suggestion
            "DateTime Arrays are not yet implemented in cudf"
```

File: python/cudf/cudf/tests/series/test_datetimelike.py
Comment (bdice) [inline python/cudf/cudf/tests/series/test_datetimelike.py:17]: I'd like to see if we can test _every_ time zone present on the system against pandas behavior for some function(s).

File: python/cudf/cudf/core/_internals/timezones.py
Comment (shwina) [reply python/cudf/cudf/core/_internals/timezones.py:160]: Ah good catch! I think that's very useful and I'll include it.

File: python/cudf/cudf/core/index.py
Comment (shwina) [reply python/cudf/cudf/core/index.py:2530]: We use `<NA>` to indicate `NaT`. Note that unlike `NaN`, `NaT` is just a "regular" null value:


```python
In [5]: cudf.Series([1, 2, None, 3], dtype="datetime64[ns]")
Out[5]: 
0    1970-01-01 00:00:00.000000001
1    1970-01-01 00:00:00.000000002
2                             <NA>
3    1970-01-01 00:00:00.000000003
dtype: datetime64[ns]

```

File: python/cudf/cudf/core/column/datetime.py
Comment (shwina) [reply python/cudf/cudf/core/column/datetime.py]: Done -- thanks for pointing this out!

File: python/cudf/cudf/core/_internals/timezones.py
Comment (shwina) [reply python/cudf/cudf/core/_internals/timezones.py:160]: Done -- `.tz_localize(None)` now converts timezone-aware timestamps to timezone-naive ones.

File: python/cudf/cudf/core/_internals/timezones.py
Comment (shwina) [reply python/cudf/cudf/core/_internals/timezones.py:101]: Fixed

**Review by mroeschke — APPROVED** [review]
Nice! My comments have been addressed

**Comment by shwina** [comment]
This PR is blocked on dropping Python 3.8 and adding Python 3.9 support in cuDF (the `zoneinfo` module was introduced in Python 3.9)

**Comment by bdice** [comment]
Waiting for #13196.

File: python/cudf/cudf/core/_internals/timezones.py
Comment (bdice) [inline python/cudf/cudf/core/_internals/timezones.py:104]:
```suggestion
    size as `data`, that respectively indicate ambiguous and
```

File: python/cudf/cudf/core/_internals/timezones.py
Comment (bdice) [inline python/cudf/cudf/core/_internals/timezones.py:118]: The docs say that "no transitions" returns `(None, None)`. How do we reconcile this?

File: python/cudf/cudf/core/column/datetime.py
Comment (bdice) [inline python/cudf/cudf/core/column/datetime.py:573]: Should this property error implementation be inherited from `DateTimeColumn`? Or if you think it's possible that CuPy might support time-zone-naive data (the parent class but not this zone-aware class), should we mention "time zone" in the error message?

File: python/cudf/cudf/core/index.py
Comment (bdice) [inline python/cudf/cudf/core/index.py:2504]: I think there's usually supposed to be a space on both sides of the colon here.

> The colon must be preceded by a space, or omitted if the type is absent.

https://numpydoc.readthedocs.io/en/latest/format.html#parameters

```suggestion
        tz : str
```

File: python/cudf/cudf/core/series.py
Comment (bdice) [inline python/cudf/cudf/core/series.py:4605]: This might be weird, but I would expect to copy the docstring from Series to Index, rather than Index to Series. I view the Series as the more fundamental object, but I don't know what precedents exist here in either class.

File: python/cudf/cudf/tests/series/test_datetimelike.py
Comment (bdice) [inline python/cudf/cudf/tests/series/test_datetimelike.py:15]: Just FYI, we ran into issues with `/etc/localtime` not existing in https://github.com/rapidsai/mambaforge-cuda/pull/16. The relation here is that I'm not 100% sure if you can assume `/usr/share/zoneinfo` is populated with data in super-minimalist Docker containers. What this means is that the result of this function could be an empty list. If you want that to raise an error/warning, you might want to add something explicit for that. I'm content with the current behavior or a warning. I might lean away from raising an error, but I'm not sure.

File: python/cudf/cudf/tests/series/test_datetimelike.py
Comment (bdice) [reply python/cudf/cudf/tests/series/test_datetimelike.py:37]: Yes, this is a good solution. I think testing all timezones for a very limited subset of functions is a helpful way to make sure we're not making bad assumptions about the timezone database structure which may change over time or by system.

File: python/cudf/cudf/tests/series/test_datetimelike.py
Comment (bdice) [inline python/cudf/cudf/tests/series/test_datetimelike.py:84]: Were we able to identify if there is an inconsistency somewhere for Grand Turk and Metlakatla? Perhaps disagreements between system-provided timezone information and some timezone database in a Python package?

Comment (bdice) [reply python/cudf/cudf/tests/series/test_datetimelike.py:84]: This also might be a candidate for skipping instead of xfailing, if we run tests with `xfail_strict`. This sounds like something where we could get an unexpected success (which would cause an error) due to Python/system software updates ...particularly if my hypothesis above is true.

**Review by bdice — COMMENTED** [review]
Mostly minor comments. I'll approve once you've taken a look and addressed some of them. This is great work -- nice job @shwina.

File: python/cudf/cudf/tests/series/test_datetimelike.py
Comment (shwina) [reply python/cudf/cudf/tests/series/test_datetimelike.py:84]: I honestly haven't gotten to the bottom of it. If we ever _do_ hit an unexpected success here, e.g., if something changes in either `zoneinfo` or `pytz`, I would love to be alerted of that! So, should we keep the `xfail`?

File: python/cudf/cudf/core/_internals/timezones.py
Comment (shwina) [reply python/cudf/cudf/core/_internals/timezones.py:118]: I changed the docs

Comment (shwina) [reply python/cudf/cudf/core/_internals/timezones.py:118]: (I wish I was the kind of programmer whose docs were more up-to-date than their code was)

File: python/cudf/cudf/core/column/datetime.py
Comment (shwina) [reply python/cudf/cudf/core/column/datetime.py]: Nope -- you're right, we should just inherit this

File: python/cudf/cudf/core/index.py
Comment (shwina) [reply python/cudf/cudf/core/index.py]: Fix'd

File: python/cudf/cudf/core/column/__init__.py
Comment (bdice) [inline python/cudf/cudf/core/column/__init__.py:26]: FYI: you can fix these "unused import" errors by defining `__all__`. If it's in `__all__`, it's "part of the API" of this module and thus flake8 sees it as used.

File: python/cudf/cudf/core/column/datetime.py
Comment (galipremsagar) [inline python/cudf/cudf/core/column/datetime.py:527]: This might need an update to places here to include `DatetimeTZColumn`:

https://github.com/rapidsai/cudf/blob/branch-23.06/python/cudf/cudf/core/series.py#L1545-L1552
https://github.com/rapidsai/cudf/blob/branch-23.06/python/cudf/cudf/core/dataframe.py#L3801-L3815
https://github.com/rapidsai/cudf/blob/branch-23.06/python/cudf/cudf/core/dataframe.py#L1702-L1713

File: python/cudf/cudf/core/column/column.py
Comment (bdice) [inline python/cudf/cudf/core/column/column.py:1559]: `ValueError` seems more natural here unless you're matching something in pandas.
```suggestion
            raise ValueError("Must specify data buffer")
```

File: python/cudf/cudf/core/series.py
Comment (bdice) [reply python/cudf/cudf/core/series.py:4605]: Sounds good. I agree with you!

File: python/cudf/cudf/tests/series/test_datetimelike.py
Comment (bdice) [reply python/cudf/cudf/tests/series/test_datetimelike.py:15]: No strong feelings. Let's leave it as-is and not overengineer it just to get a skip message.

File: python/cudf/cudf/tests/series/test_datetimelike.py
Comment (bdice) [reply python/cudf/cudf/tests/series/test_datetimelike.py:84]: Sounds good to me!

**Review by bdice — APPROVED** [review]
Great work. I had a couple small comments/replies but nothing blocking. Thanks.

File: python/cudf/cudf/core/column/datetime.py
Comment (galipremsagar) [reply python/cudf/cudf/core/column/datetime.py:527]: While we are at it, can we see if can totally remove the special casing in some those places and call `_with_type_metadata` always?

File: python/cudf/cudf/core/column/__init__.py
Comment (shwina) [reply python/cudf/cudf/core/column/__init__.py:26]: Isn't this more `explicit` though? I always thought the only "official" use for `__all__` was to define what happens when you do `from module import *`

**Review by galipremsagar — APPROVED** [review]

File: python/cudf/cudf/core/column/datetime.py
Comment (shwina) [reply python/cudf/cudf/core/column/datetime.py:527]: Done (and done!)

File: python/cudf/cudf/core/column/__init__.py
Comment (bdice) [reply python/cudf/cudf/core/column/__init__.py:26]: I think that defining your module’s public API (rather than not defining it) is the most `explicit` approach. I’m pro-`__all__` in Python libraries.

Comment (shwina) [reply python/cudf/cudf/core/column/__init__.py:26]: >  think that defining your module’s public API (rather than not defining it) is the most `explicit` approach. 

Hmm, but `__all__` doesn't do that AFAICT. Placing a name in `__all__` doesn't preclude it from appearing in e.g., tab completion options (naming it with a leading `_` does). 

Are there other tools than `flake8` that do something with `__all__`?

Comment (mroeschke) [reply python/cudf/cudf/core/column/__init__.py:26]: I think type checkers (mypy, pyright) will utilize `__all__` to see what APIs are "public"

Comment (bdice) [reply python/cudf/cudf/core/column/__init__.py:26]: Also, to your question about tab completion in IPython/etc., that's discussed here: https://github.com/ipython/ipykernel/issues/129#issuecomment-216870597

The community seems split on the issue, but some folks seem to think limiting to `__all__` by default would be good. Either way -- it seems that `__all__` is viewed as a source of truth for public APIs, and the real question is whether autocompletion should only show public APIs.

> It's configurable, with the option IPCompleter.limit_to__all__ ([config options docs](http://ipython.readthedocs.io/en/stable/config/options/terminal.html)).
>
> The reason we do this is that we often want to use IPython to poke about and debug modules, so we like to see what's really there, not just what's whitelisted as public API. We should, however, do a better job of prioritising completions, so that things in `__all__` come before things that aren't in it.

Comment (shwina) [reply python/cudf/cudf/core/column/__init__.py:26]: Thanks, both! I'm on board with defining `__all__`

Comment (shwina) [reply python/cudf/cudf/core/column/__init__.py:26]: I wonder if we can get a linter to enforce that `__all__` is defined (even if empty) always.

**Pull request merged**
